core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -128,15 +154,28 @@ public void clusterChanged(ClusterChangedEvent event) {
                     } else {
                         indexSettings = indexService.getIndexSettings();
                     }
-                    if (indicesService.canDeleteShardContent(shardId, indexSettings)) {
-                        deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
```

Comment:
I wish we somehow worked "Exists locally but can be deleted" into the name. How about `FOLDER_FOUND_CAN_DELETE` ? (it's long, I know :))